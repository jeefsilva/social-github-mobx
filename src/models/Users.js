import { types, getParent, destroy } from "mobx-state-tree";


export const User = types
  .model({
    "login": types.string,
    "avatar_url": "",
    "html_url": "",
    "name": types.optional(types.maybeNull(types.string), null),
    "blog": "",
    "location": types.optional(types.maybeNull(types.string), null),
    "public_repos": types.number,
    "followers": types.number
  })
  .actions(self => ({
    changeLogin(newLogin) {
      self.login = newLogin;
    },
    changeId(newId) {
      self.id = newId;
    },
    changeName(newName) {
      self.name = newName;
    },
    changeCompany(newCompany) {
      self.company = newCompany;
    },
    changeBlog(newBlog) {
      self.blog = newBlog;
    },
    changeLocation(newLocation) {
      self.location = newLocation;
    },
    changePublicRepos(newPublicRepo) {
      self.public_repos = newPublicRepo;
    },
    changeFollowers(newFollowers) {
      self.followers = newFollowers;
    },
    remove() {
      getParent(self, 2).remove(self)
    }
  }));

export const Users = types
  .model({
    users: types.optional(types.array(User), []),
  })
  .actions(self => ({
    add(user) {
      self.users.unshift(user);
    },
    addUsers(user) {
      self.users.push(...user)
    },
    remove(user) {
      destroy(user)
    },
  }))
  .views(self => ({
      get totalUser() {
          return self.users.length
      }
  }))
