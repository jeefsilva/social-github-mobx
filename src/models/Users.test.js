import { getSnapshot, onSnapshot, onPatch } from "mobx-state-tree";
import { Users, User } from "./Users";
import { reaction } from "mobx"

it("can create a instance of a model", () => {
  const data = User.create({
    login: "mojombo",
    id: 1,
    avatar_url: "https://avatars0.githubusercontent.com/u/1?v=4",
    html_url: "https://github.com/mojombo",
    name: "Tom Preston-Werner",
    company: "",
    blog: "http://tom.preston-werner.com",
    location: "San Francisco",
    public_repos: 61,
    followers: 21658
  });

  expect(data.company).toBe("");
  data.changeLogin("jeefsilva");
  expect(data.login).toBe("jeefsilva");
});

it("can add new users - 2", () => {
  const userList = Users.create();
  const patches = [];
  onPatch(userList, snapshot => {
    patches.push(snapshot);
  });
  userList.add({
    login: "jeefsilva",
    id: 2,
    avatar_url: "",
    html_url: "https://github.com/jeefsilva",
    name: "Jefferson Silva",
    company: "",
    blog: "http://aagenciasalvare.com.br",
    location: "São Paulo",
    public_repos: 6,
    followers: 1
  });

  userList.users[0].changeLogin("roberto");

  expect(patches).toMatchSnapshot();
});

it("can calculate the total of users", () => {
  const userList = Users.create({
    users: [
      {
        login: "roberto",
        id: 1,
        avatar_url: "",
        html_url: "https://github.com/jeefsilva",
        name: "Jefferson Silva",
        company: "",
        blog: "http://aagenciasalvare.com.br",
        location: "São Paulo",
        public_repos: 6,
        followers: 1
      },
      {
        login: "jeefsilva",
        id: 2,
        avatar_url: "",
        html_url: "https://github.com/jeefsilva",
        name: "Jefferson Silva",
        company: "",
        blog: "http://aagenciasalvare.com.br",
        location: "São Paulo",
        public_repos: 6,
        followers: 1
      }
    ]
  })

  expect(userList.totalUser).toBe(12)

  let changed = 0
  reaction(() => userList.totalUser, () => changed++)

  expect(changed).toBe(0)
  console.log(userList.totalUser)
  userList.users[0].changeLogin("oswaldo")
  expect(changed).toBe(0)
  userList.users[0].changePublicRepos(18)
  expect(changed).toBe(1)
});
