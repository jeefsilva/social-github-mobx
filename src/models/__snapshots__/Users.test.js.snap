// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can add new users - 2 1`] = `
Array [
  Object {
    "op": "add",
    "path": "/users/0",
    "value": Object {
      "avatar_url": "",
      "blog": "http://aagenciasalvare.com.br",
      "company": "",
      "followers": 1,
      "html_url": "https://github.com/jeefsilva",
      "id": 2,
      "location": "S達o Paulo",
      "login": "jeefsilva",
      "name": "Jefferson Silva",
      "public_repos": 6,
    },
  },
  Object {
    "op": "replace",
    "path": "/users/0/login",
    "value": "roberto",
  },
]
`;

exports[`can add new users 1`] = `
Object {
  "users": Array [
    Object {
      "avatar_url": "",
      "blog": "http://aagenciasalvare.com.br",
      "company": "",
      "followers": 1,
      "html_url": "https://github.com/jeefsilva",
      "id": 2,
      "location": "S達o Paulo",
      "login": "roberto",
      "name": "Jefferson Silva",
      "public_repos": 6,
    },
  ],
}
`;

exports[`can add new users 2`] = `
Array [
  Object {
    "users": Array [
      Object {
        "avatar_url": "",
        "blog": "http://aagenciasalvare.com.br",
        "company": "",
        "followers": 1,
        "html_url": "https://github.com/jeefsilva",
        "id": 2,
        "location": "S達o Paulo",
        "login": "jeefsilva",
        "name": "Jefferson Silva",
        "public_repos": 6,
      },
    ],
  },
  Object {
    "users": Array [
      Object {
        "avatar_url": "",
        "blog": "http://aagenciasalvare.com.br",
        "company": "",
        "followers": 1,
        "html_url": "https://github.com/jeefsilva",
        "id": 2,
        "location": "S達o Paulo",
        "login": "roberto",
        "name": "Jefferson Silva",
        "public_repos": 6,
      },
    ],
  },
]
`;
